_examples/basic/2-realtime-feed/producer/main.go

```diff
@@ -0,0 +1,92 @@
+package main
+
+import (
+	"encoding/json"
+	"math/rand"
+	"time"
+
+	"github.com/brianvoe/gofakeit"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill-kafka/pkg/kafka"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/ThreeDotsLabs/watermill/message/router/middleware"
+)
+
+var (
+	brokers = []string{"kafka:9092"}
+
+	messagesToPublish = 10000
+	workers           = 25
+)
+
+func main() {
+	logger := watermill.NewStdLogger(false, false)
+	logger.Info("Starting the producer", watermill.LogFields{})
+
+	publisher, err := kafka.NewPublisher(brokers, kafka.DefaultMarshaler{}, nil, logger)
+	if err != nil {
+		panic(err)
+	}
+	defer publisher.Close()
+
+	msgPublished := make(chan struct{})
+	allMessagesPublished := make(chan struct{})
+
+	go func() {
+		for range msgPublished {
+			messagesToPublish--
+
+			if messagesToPublish%1000 == 0 {
+				logger.Info("messages left", watermill.LogFields{"count": messagesToPublish})
+			}
+			if messagesToPublish == 0 {
+				allMessagesPublished <- struct{}{}
+			}
+		}
+	}()
+
+	for num := 0; num < workers; num++ {
+		go func() {
+			for messagesToPublish > 0 {
```

Comment:
Isn't this going to race with the write at 38? it's not synced with the other goroutine.

Reply:
Nice catch. 👍 Reworked to distribute work by channels in 3330bd4d8c107b3fdcbd5bc4f25369adedd17f48

Reply:
This could perhaps be made even clearer if `chan message.Message` were used instead of the struct channels?
I'll try to propose my version in a separate PR

Reply:
https://github.com/ThreeDotsLabs/watermill/pull/121/files

Reply:
I made some manual testing and it seems that all the messages are published. The channel is buffered to the number of workers, so `close` shouldn't be hit before all messages are consumed.

I'm not 100% sure that we are losing no message here, but I don't know if it's crucial in an example like this

Reply:
#121 merged.